20220714-lookup-api.md

```diff
@@ -0,0 +1,457 @@
+  
+For production usage we should document recommendations for these configuration limits so as to protect the server from memory exhaustion or costly Lookup queries.
+
+## Server Changes
+* Add implementations of the API changes described in [Changes to the OpenFGA API](#changes-to-the-openfga-api) in the server. This includes implementations:
+```
+func (s *Server) StreamedLookup(
+    req *openfgapb.LookupRequest,
+    srv openfgapb.OpenFGAService_StreamedLookupServer,
+) error {...}
+
+func (s *Server) Lookup(
+    ctx context.Context,
+    req *openfgapb.LookupRequest,
+) (*openfgapb.LookupResponse, error) {...}
+```
+
+Sample code implementing these proposed changes can be found here:
+https://github.com/openfga/openfga/tree/lookup-poc (server changes) which depends on https://github.com/openfga/api/tree/lookup-api (protobuf API changes)
+
+### Storage Interface Changes
+The [`OpenFGADatastore`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L132) interface will need to be expanded or modified to support queries for all objects of a particular object type (within a store). Today our datastore/storage interface only has the [`Read`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L51) method, but it requires a storeID and tuple key specifying either the `object` or `user` field (or both). The Lookup work will need a method that supports only providing the storeID and object `type`.
+
+We should modify the [`TupleBackend`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L47) interface with a new method `QueryRelationshipTuples` and it's signature could look like:
+
+```
+QueryRelationships(
+  ctx context.Context,
+  storeID string,
+  filter *RelationshipFilter,
+  opts ...QueryOptionsOption,
+) (TupleIterator, error)
+
+type RelationshipFilter struct {
+  ObjectType         string
+  OptionalObjectID   string
+  OptionalRelation   string
+  OptionalUserFilter string
+}
+
+type QueryOptions struct {
+  Limit    *uint64
+  Usersets []string
+}
+
+type QueryOptionsOption func(*QueryOptions)
+```
+
+This approach will allow us to query relationship tuples using any provided filters, and could serve as a more general replacement for the existing Read method.
+
+# Migration
+[migration]: #migration
+
+No migrations or breaking changes are necessary to support the Lookup API endpoints described herein.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Potential arguments for not introducing the Lookup API include:
+* Without proper limitations clients could impact the whole system by issuing Lookups on large collections.
+* Increases the OpenFGA API surface for a very niche use case for Search with Permissions on small object collections.
```

Comment:
* encourages users to use Lookup over other methods that may be better applied for their use-case (Search then Check or Local Index)

Reply:
Added this to the list. Marking as resolved.